[WebRTC] Support for events and stats aggregation

philn opened the pull request:
#### b36717dc42b180575f2a695f04703b2765c1d28e
<pre>
[WebRTC] Support for events and stats aggregation
<a href="https://bugs.webkit.org/show_bug.cgi?id=279433">https://bugs.webkit.org/show_bug.cgi?id=279433</a>
&lt;<a href="https://rdar.apple.com/problem/136139817">rdar://problem/136139817</a>&gt;

Reviewed by Eric Carlson.

This patch instruments the LibWebRTC and GstWebRTC backends with additional logging that can be
gathered and processed by the PeerConnectionBackend. In this first version, for WPE and GTK ports,
the collected logs are written to a JSON file, if the WEBKIT_WEBRTC_JSON_EVENTS_FILE environment
variable is set.

gatherLogs is also now supported by the GStreamer backend, by using a custom GST_DEBUG log handler.

Future versions could send the JSON stream to the WebInspector or a custom URI protocol handler in
the UIProcess, for live graphing purposes.

* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/wtf/Logger.h:
(WTF::Logger::Observer::handleLogMessage):
(WTF::Logger::logAlways const):
(WTF::Logger::error const):
(WTF::Logger::warning const):
(WTF::Logger::info const):
(WTF::Logger::debug const):
(WTF::Logger::logAlwaysVerbose const):
(WTF::Logger::errorVerbose const):
(WTF::Logger::warningVerbose const):
(WTF::Logger::infoVerbose const):
(WTF::Logger::debugVerbose const):
(WTF::Logger::willLog const):
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
(WebCore::PeerConnectionBackend::PeerConnectionBackend):
(WebCore::PeerConnectionBackend::~PeerConnectionBackend):
(WebCore::PeerConnectionBackend::handleLogMessage):
(WebCore::PeerConnectionBackend::createOfferSucceeded):
(WebCore::PeerConnectionBackend::createOfferFailed):
(WebCore::PeerConnectionBackend::createAnswerSucceeded):
(WebCore::PeerConnectionBackend::createAnswerFailed):
(WebCore::PeerConnectionBackend::setLocalDescriptionSucceeded):
(WebCore::PeerConnectionBackend::setLocalDescriptionFailed):
(WebCore::PeerConnectionBackend::setRemoteDescriptionSucceeded):
(WebCore::PeerConnectionBackend::setRemoteDescriptionFailed):
(WebCore::PeerConnectionBackend::newICECandidate):
(WebCore::PeerConnectionBackend::newDataChannel):
(WebCore::PeerConnectionBackend::generateJSONLogEvent):
(WebCore::PeerConnectionBackend::emitJSONLogEvent):
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.h:
(WebCore::PeerConnectionBackend::isJSONLogStreamingEnabled const):
(WebCore::PeerConnectionBackend::setJSONLogStreamingEnabled):
* Source/WebCore/Modules/mediastream/RTCController.cpp:
(WebCore::RTCController::startGatheringLogs):
(WebCore::RTCController::stopGatheringLogs):
(WebCore::RTCController::stopLoggingWebRTCLogs):
(WebCore::RTCController::stopLoggingLibWebRTCLogs): Deleted.
* Source/WebCore/Modules/mediastream/RTCController.h:
* Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
(WebCore::RTCPeerConnection::setLocalDescription):
(WebCore::RTCPeerConnection::setRemoteDescription):
(WebCore::RTCPeerConnection::addIceCandidate):
(WebCore::RTCPeerConnection::dispatchEvent):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::GStreamerMediaEndpoint):
(WebCore::RTCStatsLogger::toJSONString const):
(WebCore::GStreamerMediaEndpoint::processStatsItem):
(WebCore::GStreamerMediaEndpoint::statsLogInterval const):
(WebCore::GStreamerMediaEndpoint::startRTCLogs):
(WebCore::GStreamerMediaEndpoint::stopRTCLogs):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::GStreamerPeerConnectionBackend::startGatheringStatLogs):
(WebCore::GStreamerPeerConnectionBackend::stopGatheringStatLogs):
(WebCore::GStreamerPeerConnectionBackend::provideStatLogs):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::LibWebRTCMediaEndpoint::LibWebRTCMediaEndpoint):
(WebCore::LibWebRTCMediaEndpoint::OnStatsDelivered):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.h:
* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::setLogger):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerWebRTCLogSink.cpp: Added.
(WebCore::GStreamerWebRTCLogSink::GStreamerWebRTCLogSink):
(WebCore::toWebRTCLogLevel):
(WebCore::GStreamerWebRTCLogSink::start):
(WebCore::GStreamerWebRTCLogSink::stop):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerWebRTCLogSink.h: Added.

Canonical link: <a href="https://commits.webkit.org/287090@main">https://commits.webkit.org/287090@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d85dfb91c46b329f79866379f8dbf478c49852e0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78301 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57347 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31683 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82962 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29567 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80434 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66498 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5628 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/61324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/19244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81368 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/51322 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/68576 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/41640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48669 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24974 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27904 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71448 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/69761 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25340 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84327 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77541 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5668 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3830 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/69547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5827 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67267 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/68803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17150 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12812 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11111 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/99848 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5615 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/8363 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21811 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5607 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9041 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7393 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
```diff
@@ -129,10 +139,47 @@ void PeerConnectionBackend::createOffer(RTCOfferOptions&& options, CreateCallbac
+    if (!document)
+        return false;
+
+    auto page = document->page();
```
This triggers an assert because can be called from a non-main thread. I'll see what i can do about this...

```
STDERR: ASSERTION FAILED: canSafelyBeUsed()
STDERR: /app/webkit/WebKitBuild/WPE/Release/WTF/Headers/wtf/WeakPtr.h(118) : T* WTF::WeakPtr< <template-parameter-1-1>, WeakPtrImpl, <template-parameter-1-3> >::operator->() const [with T = WebCore::LocalFrame; WeakPtrImpl = WTF::DefaultWeakPtrImpl; PtrTraits = WTF::RawPtrTraits<WTF::DefaultWeakPtrImpl>]
STDERR: 1   0x7f10af2ff821 WebCore::Document::page() const
STDERR: 2   0x7f10b2b0e627 WebCore::PeerConnectionBackend::handleLogMessage(WTFLogChannel const&, WTFLogLevel, WTF::Vector<WTF::JSONLogValue, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&&)
STDERR: 3   0x7f10b43db16d WebCore::RealtimeMediaSource::videoFrameAvailable(WebCore::VideoFrame&, WebCore::VideoFrameTimeMetadata)
STDERR: 4   0x7f10b43f2451 WebCore::RealtimeVideoCaptureSource::dispatchVideoFrameToObservers(WebCore::VideoFrame&, WebCore::VideoFrameTimeMetadata)
STDERR: 5   0x7f10b441c33f WTF::Detail::CallableWrapper<WebCore::MockRealtimeVideoSourceGStreamer::MockRealtimeVideoSourceGStreamer(WTF::String&&, WTF::AtomString&&, WebCore::MediaDeviceHashSalts&&, std::optional<WTF::ObjectIdentifierGeneric<WebCore::PageIdentifierType, WTF::ObjectIdentifierMainThreadAccessTraits<unsigned long>, unsigned long> >)::{lambda(auto:1&&)#1}, void, WTF::Ref<WebCore::VideoFrameGStreamer, WTF::RawPtrTraits<WebCore::VideoFrameGStreamer>, WTF::DefaultRefDerefTraits<WebCore::VideoFrameGStreamer> >&&>::call(WTF::Ref<WebCore::VideoFrameGStreamer, WTF::RawPtrTraits<WebCore::VideoFrameGStreamer>, WTF::DefaultRefDerefTraits<WebCore::VideoFrameGStreamer> >&&)
STDERR: 6   0x7f10b440be38 WebCore::GStreamerVideoCapturer::setSinkVideoFrameCallback(WTF::Function<void (WTF::Ref<WebCore::VideoFrameGStreamer, WTF::RawPtrTraits<WebCore::VideoFrameGStreamer>, WTF::DefaultRefDerefTraits<WebCore::VideoFrameGStreamer> >&&)>&&)::{lambda(WebCore::GStreamerVideoCapturer*, _GstElement*)#1}::_FUN(WebCore::GStreamerVideoCapturer*, _GstElement*)
STDERR: 7   0x7f10a43f9056 ffi_call_unix64
STDERR: 8   0x7f10a43f7b5d ffi_call_int
STDERR: 9   0x7f10a43f8313 ffi_call
STDERR: 10  0x7f10a8fccb43 g_cclosure_marshal_generic_va
STDERR: 11  0x7f10a8fe6a83 g_signal_emit_valist
STDERR: 12  0x7f10a8fe6c03 g_signal_emit
STDERR: 13  0x7f10a88ddbab gst_app_sink_render_common.isra.0
STDERR: 14  0x7f10a647c4e3 gst_base_sink_chain_unlocked.isra.0
STDERR: 15  0x7f10a644901e gst_base_sink_chain_main
STDERR: 16  0x7f10a972cb6c gst_pad_chain_data_unchecked
STDERR: 17  0x7f10a972ff0e gst_pad_push_data
STDERR: 18  0x7f10a9730544 gst_pad_push
STDERR: 19  0x7f100821f0e9 gst_queue_loop
STDERR: 20  0x7f10a97647d4 gst_task_func
STDERR: 21  0x7f10a99464f2 g_thread_pool_thread_proxy.lto_priv.0
STDERR: 22  0x7f10a9943862 g_thread_proxy
STDERR: 23  0x7f10a90a1e39 start_thread
STDERR: 24  0x7f10a91298c4 clone
```

philn commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
```diff
@@ -826,6 +827,13 @@ class RTCStatsLogger {
 
 void LibWebRTCMediaEndpoint::OnStatsDelivered(const rtc::scoped_refptr<const webrtc::RTCStatsReport>& report)
 {
+    auto backend = protectedPeerConnectionBackend();
```
```
ASSERTION FAILED: canSafelyBeUsed()
/Volumes/Data/worker/macOS-Sequoia-Debug-Build-EWS/build/WebKitBuild/Debug/usr/local/include/wtf/WeakRef.h(111) : T *WTF::WeakRef<WebCore::RTCPeerConnection, WebCore::WeakPtrImplWithEventTargetData>::operator->() const [T = WebCore::RTCPeerConnection, WeakPtrImpl = WebCore::WeakPtrImplWithEventTargetData]
1   0x3031a1d78 WTF::WeakRef<WebCore::RTCPeerConnection, WebCore::WeakPtrImplWithEventTargetData>::operator->() const
2   0x3031a54e8 WebCore::PeerConnectionBackend::ref() const
3   0x30328cd2c WTF::DefaultRefDerefTraits<WebCore::LibWebRTCPeerConnectionBackend>::ref(WebCore::LibWebRTCPeerConnectionBackend&)
4   0x30328ccfc WTF::Ref<WebCore::LibWebRTCPeerConnectionBackend, WTF::RawPtrTraits<WebCore::LibWebRTCPeerConnectionBackend>, WTF::DefaultRefDerefTraits<WebCore::LibWebRTCPeerConnectionBackend>>::Ref(WebCore::LibWebRTCPeerConnectionBackend&)
5   0x303251ba8 WTF::Ref<WebCore::LibWebRTCPeerConnectionBackend, WTF::RawPtrTraits<WebCore::LibWebRTCPeerConnectionBackend>, WTF::DefaultRefDerefTraits<WebCore::LibWebRTCPeerConnectionBackend>>::Ref(WebCore::LibWebRTCPeerConnectionBackend&)
6   0x3032525b0 WebCore::LibWebRTCMediaEndpoint::protectedPeerConnectionBackend() const
7   0x3032572e0 WebCore::LibWebRTCMediaEndpoint::OnStatsDelivered(webrtc::scoped_refptr<webrtc::RTCStatsReport const> const&)
8   0x108616818 webrtc::RTCStatsCollector::DeliverCachedReport(webrtc::scoped_refptr<webrtc::RTCStatsReport const>, std::__1::vector<webrtc::RTCStatsCollector::RequestInfo, std::__1::allocator<webrtc::RTCStatsCollector::RequestInfo>>)
9   0x108617f4c webrtc::RTCStatsCollector::MergeNetworkReport_s()
10  0x108639db8 webrtc::RTCStatsCollector::ProducePartialResultsOnNetworkThread(webrtc::Timestamp, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>)::$_1::operator()() const
11  0x108639d90 decltype(std::declval<webrtc::RTCStatsCollector::ProducePartialResultsOnNetworkThread(webrtc::Timestamp, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>)::$_1>()()) std::__1::__invoke[abi:sn180100]<webrtc::RTCStatsCollector::ProducePartialResultsOnNetworkThread(webrtc::Timestamp, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>)::$_1>(webrtc::RTCStatsCollector::ProducePartialResultsOnNetworkThread(webrtc::Timestamp, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>)::$_1&&)
12  0x108639d6c std::__1::invoke_result<webrtc::RTCStatsCollector::ProducePartialResultsOnNetworkThread(webrtc::Timestamp, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>)::$_1>::type std::__1::invoke[abi:sn180100]<webrtc::RTCStatsCollector::ProducePartialResultsOnNetworkThread(webrtc::Timestamp, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>)::$_1>(webrtc::RTCStatsCollector::ProducePartialResultsOnNetworkThread(webrtc::Timestamp, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>)::$_1&&)
13  0x108639d10 void absl::internal_any_invocable::InvokeR<void, webrtc::RTCStatsCollector::ProducePartialResultsOnNetworkThread(webrtc::Timestamp, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>)::$_1, void>(webrtc::RTCStatsCollector::ProducePartialResultsOnNetworkThread(webrtc::Timestamp, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>)::$_1&&)
14  0x108639c74 void absl::internal_any_invocable::LocalInvoker<false, void, webrtc::RTCStatsCollector::ProducePartialResultsOnNetworkThread(webrtc::Timestamp, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>)::$_1&&>(absl::internal_any_invocable::TypeErasedState*)
15  0x107a5b0b8 absl::internal_any_invocable::Impl<void () &&>::operator()() &&
16  0x108ae1668 rtc::Thread::Dispatch(absl::AnyInvocable<void () &&>)
17  0x108adefdc rtc::Thread::ProcessMessages(int)
18  0x108ae2a64 rtc::Thread::Run()
19  0x108ae26bc rtc::Thread::PreRun(void*)
20  0x19813b2e4 _pthread_start
21  0x1981360fc thread_start
com.apple.WebKit.WebContent.Development terminated (pid 9576) for reason: crash
LEAK: 1 WebPageProxy
```

youennf commented on Source/WTF/wtf/Logger.h:
```diff
@@ -240,14 +241,26 @@ class Logger : public ThreadSafeRefCounted<Logger> {
+
+            Locker locker { AdoptLock, observerLock() };
+            for (Observer& observer : observers()) {
+                if (observer.handleLogMessage(channel, level, { ConsoleLogValue<Argument>::toValue(arguments)... }))
```
Why not doing the work in `didLogMessage` instead?

philn replied:
Because then this would become dependent on having release logs enabled at runtime (on Linux ports by setting this env var: WEBKIT_DEBUG=WebRTC=enabled) and thus the logs would also be sent to systemd journal.

The `didLogMessage` handler is not called if nothing was logged, that's why I added a new method.

youennf replied:
OK. Do we need `handleLogMessage` to return a boolean?
It seems it could only be a passive listener.

youennf commented on Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
```diff
@@ -145,7 +197,7 @@ void PeerConnectionBackend::createOfferSucceeded(String&& sdp)
 {
     ASSERT(isMainThread());
-    ALWAYS_LOG(LOGIDENTIFIER, "Create offer failed:", exception.message());
+    ALWAYS_LOG(LOGIDENTIFIER, makeString("Create offer failed:"_s, exception.message()));
```
It is less efficient to have the makeString call if we do not log. I would tend to keep it as is.

youennf commented on Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
```diff
@@ -568,18 +623,20 @@ void PeerConnectionBackend::newICECandidate(String&& sdp, String&& mid, unsigned
 
         auto channel = RTCDataChannel::create(*peerConnection->document(), channelHandler.moveToUniquePtr(), WTFMove(label), WTFMove(channelInit), RTCDataChannelState::Open);
+        ALWAYS_LOG(LOGIDENTIFIER, makeString("Dispatching data-channel event for channel "_s, channel->label()));
         peerConnection->dispatchEvent(RTCDataChannelEvent::create(eventNames().datachannelEvent, Event::CanBubble::No, Event::IsCancelable::No, Ref { channel }));
```
I would add a peer connection helper routine, this way no need to capture `this`.

philn replied:
I see `PeerConnectionBackend::newICECandidate()` also capturing this... I can add a routine for that one too?

youennf commented on Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
```diff
@@ -707,6 +764,41 @@ void PeerConnectionBackend::deref() const
+    if (!m_isJSONLogStreamingEnabled)
+        return;
+
+    m_logFile->println(WTFMove(event));
```
This is WPE/GTK specific, maybe we should conditionally compile.
For macOS/iOS, we would probably prefer to call the callback given in `CommandLineAPIHost::gatherRTCLogs`.

youennf commented on Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
```diff
@@ -104,21 +105,134 @@ std::optional<RTCRtpCapabilities> PeerConnectionBackend::senderCapabilities(Scri
+
+JSONFileHandler& jsonFileHandler(const String& path)
+{
+    static NeverDestroyed<JSONFileHandler> sharedInstance(path);
```
What happens if `jsonFileHandler ` is called with different `path` values?
Would it be simpler to not pass the parameter and use `Singleton` as well for aligning with coding style.

philn replied:
I'm not sure what you mean, the path comes from an env var, env vars are global within the same process, the same env var can't have different values in that context.

philn replied:
`path` argument removed, the getenv is done within the singleton accessor now.

youennf commented on Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
```diff
@@ -104,21 +105,134 @@ std::optional<RTCRtpCapabilities> PeerConnectionBackend::senderCapabilities(Scri
 #if !RELEASE_LOG_DISABLED
     , m_logger(peerConnection.logger())
     , m_logIdentifier(peerConnection.logIdentifier())
+    , m_logIdentifierString(makeString(hex(m_logIdentifier)))
```
How about lazy initializng it?

youennf commented on Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
```diff
@@ -104,21 +105,134 @@ std::optional<RTCRtpCapabilities> PeerConnectionBackend::senderCapabilities(Scri
+        payloadValue = values[2].value.span8();
+
+    auto event = generateJSONLogEvent(MessageLogEvent { values[1].value, WTFMove(payloadValue) }, false);
+    emitJSONLogEvent(WTFMove(event));
```
Why not passing a MessageLogEvent struct instead of a string?
And maybe an enumeration value instead of a bool?

philn replied:
`provideStatsLogs()` takes a String.

youennf commented on Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
```diff
@@ -104,21 +105,134 @@ std::optional<RTCRtpCapabilities> PeerConnectionBackend::senderCapabilities(Scri
+{
+#if PLATFORM(WPE) || PLATFORM(GTK)
+    auto name = StringView::fromLatin1(channel.name);
+    if (name != "WebRTC"_s)
```
Would `channel != &LogWebRTC` work?
Or are we receiving WK2 logs?

philn replied:
Comparing to `&LogWebRTC` is not working because of constness of the channel.

youennf commented on Source/WebCore/Modules/mediastream/PeerConnectionBackend.h:
```diff
@@ -168,6 +169,8 @@ class PeerConnectionBackend
     ASCIILiteral logClassName() const override { return "PeerConnectionBackend"_s; }
     WTFLogChannel& logChannel() const final;
+    void didLogMessage(const WTFLogChannel&, WTFLogLevel, Vector<JSONLogValue>&&) final { }
+    bool handleLogMessage(const WTFLogChannel&, WTFLogLevel, Vector<JSONLogValue>&&) final;
```
Will an observer implement both methods?
Maybe this should be a different observer class?
The benefit would be that, in most cases, the additional observer map would be empty and this would reduce the runtime cost.

philn replied:
The default impl of `handleLogMessage()` returns false, i don't think we need a new observer only for this.

youennf commented on Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
```diff
@@ -974,10 +975,23 @@ void RTCPeerConnection::scheduleEvent(Ref<Event>&& event)
 void RTCPeerConnection::dispatchEvent(Event& event)
 {
-    INFO_LOG(LOGIDENTIFIER, "dispatching '", event.type(), "'");
+    INFO_LOG(LOGIDENTIFIER, makeString("dispatching '"_s, event.type(), '\''));
```
Can we not call makeString?

youennf commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
```diff
@@ -826,6 +827,18 @@ class RTCStatsLogger {
 void LibWebRTCMediaEndpoint::OnStatsDelivered(const rtc::scoped_refptr<const webrtc::RTCStatsReport>& report)
 {
+#if PLATFORM(WPE) || PLATFORM(GTK)
+    callOnMainThread([protectedThis = Ref { *this }, this, report] {
```
Why not reusing the next callOnMainThread call site a few lines below?

youennf commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.h:
```diff
@@ -213,6 +213,8 @@ class LibWebRTCMediaEndpoint
     bool m_isGatheringRTCLogs { false };
     bool m_shouldIgnoreNegotiationNeededSignal { false };
+
+    uintptr_t m_peerConnectionIdentifier;
```
Where is it used?

philn replied:
nowhere, removing.

youennf commented on Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
```diff
@@ -1491,7 +1491,7 @@ void RealtimeMediaSource::setLogger(const Logger& newLogger, uint64_t newLogIden
     m_logger = &newLogger;
     m_logIdentifier = newLogIdentifier;
-    ALWAYS_LOG(LOGIDENTIFIER, m_type, ", ", name(), ", ", m_hashedID, ", ", m_ephemeralHashedID);
+    ALWAYS_LOG(LOGIDENTIFIER, makeString(m_type, ", "_s, name(), ", "_s, m_hashedID, ", "_s, m_ephemeralHashedID));
```
Ditto, why makeString here?
This makes things less efficient when not logging.

youennf commented on Source/WTF/wtf/Logger.h:
```diff
@@ -240,14 +241,24 @@ class Logger : public ThreadSafeRefCounted<Logger> {
+
+            Locker locker { AdoptLock, observerLock() };
+            for (Observer& observer : observers())
+                observer.handleLogMessage(channel, level, { ConsoleLogValue<Argument>::toValue(arguments)... });
```
We do not want to construct a Vector except if handleLogMessage is actually implemented.
Otherwise we create a Vector for no good reason.
Hence why a separate observer list might be better, or some boolean state telling whether to call handleLogMessage.

eric-carlson commented on Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
```diff
@@ -116,9 +164,82 @@ PeerConnectionBackend::PeerConnectionBackend(RTCPeerConnection& peerConnection)
+    if (identifier != m_logIdentifierString)
+        return;
+
+    if (!isJSONLogStreamingEnabled())
```
Why not check this before parsing the identifier?

philn replied:
Indeed, it can be done earlier :)

philn committed:
[WebRTC] Support for events and stats aggregation
https://bugs.webkit.org/show_bug.cgi?id=279433
<rdar://problem/136139817>

Reviewed by Eric Carlson.

This patch instruments the LibWebRTC and GstWebRTC backends with additional logging that can be
gathered and processed by the PeerConnectionBackend. In this first version, for WPE and GTK ports,
the collected logs are written to a JSON file, if the WEBKIT_WEBRTC_JSON_EVENTS_FILE environment
variable is set.

gatherLogs is also now supported by the GStreamer backend, by using a custom GST_DEBUG log handler.

Future versions could send the JSON stream to the WebInspector or a custom URI protocol handler in
the UIProcess, for live graphing purposes.

* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/wtf/Logger.h:
(WTF::Logger::Observer::handleLogMessage):
(WTF::Logger::logAlways const):
(WTF::Logger::error const):
(WTF::Logger::warning const):
(WTF::Logger::info const):
(WTF::Logger::debug const):
(WTF::Logger::logAlwaysVerbose const):
(WTF::Logger::errorVerbose const):
(WTF::Logger::warningVerbose const):
(WTF::Logger::infoVerbose const):
(WTF::Logger::debugVerbose const):
(WTF::Logger::willLog const):
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
(WebCore::PeerConnectionBackend::PeerConnectionBackend):
(WebCore::PeerConnectionBackend::~PeerConnectionBackend):
(WebCore::PeerConnectionBackend::handleLogMessage):
(WebCore::PeerConnectionBackend::createOfferSucceeded):
(WebCore::PeerConnectionBackend::createOfferFailed):
(WebCore::PeerConnectionBackend::createAnswerSucceeded):
(WebCore::PeerConnectionBackend::createAnswerFailed):
(WebCore::PeerConnectionBackend::setLocalDescriptionSucceeded):
(WebCore::PeerConnectionBackend::setLocalDescriptionFailed):
(WebCore::PeerConnectionBackend::setRemoteDescriptionSucceeded):
(WebCore::PeerConnectionBackend::setRemoteDescriptionFailed):
(WebCore::PeerConnectionBackend::newICECandidate):
(WebCore::PeerConnectionBackend::newDataChannel):
(WebCore::PeerConnectionBackend::generateJSONLogEvent):
(WebCore::PeerConnectionBackend::emitJSONLogEvent):
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.h:
(WebCore::PeerConnectionBackend::isJSONLogStreamingEnabled const):
(WebCore::PeerConnectionBackend::setJSONLogStreamingEnabled):
* Source/WebCore/Modules/mediastream/RTCController.cpp:
(WebCore::RTCController::startGatheringLogs):
(WebCore::RTCController::stopGatheringLogs):
(WebCore::RTCController::stopLoggingWebRTCLogs):
(WebCore::RTCController::stopLoggingLibWebRTCLogs): Deleted.
* Source/WebCore/Modules/mediastream/RTCController.h:
* Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
(WebCore::RTCPeerConnection::setLocalDescription):
(WebCore::RTCPeerConnection::setRemoteDescription):
(WebCore::RTCPeerConnection::addIceCandidate):
(WebCore::RTCPeerConnection::dispatchEvent):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::GStreamerMediaEndpoint):
(WebCore::RTCStatsLogger::toJSONString const):
(WebCore::GStreamerMediaEndpoint::processStatsItem):
(WebCore::GStreamerMediaEndpoint::statsLogInterval const):
(WebCore::GStreamerMediaEndpoint::startRTCLogs):
(WebCore::GStreamerMediaEndpoint::stopRTCLogs):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::GStreamerPeerConnectionBackend::startGatheringStatLogs):
(WebCore::GStreamerPeerConnectionBackend::stopGatheringStatLogs):
(WebCore::GStreamerPeerConnectionBackend::provideStatLogs):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::LibWebRTCMediaEndpoint::LibWebRTCMediaEndpoint):
(WebCore::LibWebRTCMediaEndpoint::OnStatsDelivered):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.h:
* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::setLogger):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerWebRTCLogSink.cpp: Added.
(WebCore::GStreamerWebRTCLogSink::GStreamerWebRTCLogSink):
(WebCore::toWebRTCLogLevel):
(WebCore::GStreamerWebRTCLogSink::start):
(WebCore::GStreamerWebRTCLogSink::stop):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerWebRTCLogSink.h: Added.

Canonical link: https://commits.webkit.org/287090@main
